src/porepy/numerics/nonlinear/line_search.py

```diff
@@ -0,0 +1,697 @@
+
+        # Number of sampling points in the interval [0, 1]. If there is a substantial
+        # risk of local minima, this number should be increased.
+        num_steps = self.params.get("residual_line_search_num_steps", 5)
+        alpha = self.recursive_weight_from_sampling(
+            0,
+            1,
+            f_terminate,
+            objective_function,
+            f_0,
+            f_1,
+            num_steps=num_steps,
+            step_size_tolerance=interval_size,
+        )
+        # Safeguard against zero weights.
+        return np.maximum(alpha, self.min_line_search_weight) * np.ones_like(dx)
+
+    def recursive_weight_from_sampling(
+        self,
+        a: float,
+        b: float,
+        condition_function: Callable[[Sequence], bool],
+        function: Callable,
+        f_a: Optional[float] = None,
+        f_b: Optional[float] = None,
+        num_steps: Optional[int] = 5,
+        step_size_tolerance: Optional[float] = 1e-1,
+    ) -> float:
+        """Recursive function for finding a weight satisfying a condition.
+
+        The function is based on sampling the function in the interval [a, b] and
+        recursively narrowing down the interval until the interval is smaller than the
+        tolerance and the condition is satisfied. It returns the smallest tested value
+        not satisfying the condition.
+
+        Parameters:
+            a, b: The interval.
+            condition_function: The condition function. It takes a sequence of line
+                search evaluations as argument and returns True if the condition is
+                satisfied, indicating that the recursion should be terminated. The
+                condition function should be chosen such that it is compatible with the
+                property of the function to be tested (monotonicity, convexity, etc.).
+            function: The function to be tested. Returns a scalar or vector, must be
+                compatible with the condition function's parameter.
+            f_a: The value of the function at a. If not given, it is computed.
+            f_b: The value of the function at b. If not given, it is computed.
+            num_steps: The number of sampling points in the interval [a, b].
+            step_size_tolerance: The tolerance for the step size. If the step size is
+                smaller than this, the recursion is terminated.
+
+        Returns:
+            The smallest tested value not satisfying the condition.
+
+        """
+        # Initialize the lower bound of the interval and the function value at the lower
+        # bound.
+        x_l = a
+        if f_a is None:
+            f_l = function(a)
+        else:
+            f_l = f_a
+        terminate_condition = False
+        sampling_points = np.linspace(a, b, num_steps)
+        step_size = (b - a) / (num_steps - 1)
+        f_vals = [f_l]
+
+        # x_h is the upper bound of the interval in the next iteration.
```

Comment:
Thanks for improving the documentation. I made a few suggestions for more details; if these are roughly correct (though feel free to improve formulations), I am happy with the level of documentation - and believe I may be so if I need to look at this in six months time, say.